ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsMaterializedSessionTableIQv2.java

```diff
@@ -0,0 +1,115 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.streams.materialization.ks;
+
+import static org.apache.kafka.streams.query.StateQueryRequest.inStore;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+import com.google.common.collect.Range;
+import io.confluent.ksql.GenericKey;
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.execution.streams.materialization.MaterializationException;
+import io.confluent.ksql.execution.streams.materialization.MaterializedWindowedTable;
+import io.confluent.ksql.execution.streams.materialization.WindowedRow;
+import java.time.Instant;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.Window;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.query.StateQueryRequest;
+import org.apache.kafka.streams.query.StateQueryResult;
+import org.apache.kafka.streams.query.WindowRangeQuery;
+import org.apache.kafka.streams.state.KeyValueIterator;
+
+/**
+ * Kafka Streams impl of {@link MaterializedWindowedTable}.
+ */
+class KsMaterializedSessionTableIQv2 implements MaterializedWindowedTable {
+
+  private final KsStateStore stateStore;
+
+  KsMaterializedSessionTableIQv2(final KsStateStore store) {
+    this.stateStore = Objects.requireNonNull(store, "store");
+  }
+
+  @Override
+  public List<WindowedRow> get(
+      final GenericKey key,
+      final int partition,
+      final Range<Instant> windowStart,
+      final Range<Instant> windowEnd
+  ) {
+    try {
+      return findSession(key, windowStart, windowEnd);
+    } catch (final Exception e) {
+      throw new MaterializationException("Failed to get value from materialized table", e);
+    }
+  }
+
+  @Override
+  public Iterator<WindowedRow> get(final int partition, final Range<Instant> windowStartBounds,
+      final Range<Instant> windowEndBounds) {
+    throw new MaterializationException("Table scan unsupported on session tables");
+  }
+
+  private List<WindowedRow> findSession(
+      final GenericKey key,
+      final Range<Instant> windowStart,
+      final Range<Instant> windowEnd
+  ) {
+
+    final WindowRangeQuery<GenericKey, GenericRow> query = WindowRangeQuery.withKey(key);
+    final StateQueryRequest<KeyValueIterator<Windowed<GenericKey>, GenericRow>> request =
+        inStore(stateStore.getStateStoreName()).withQuery(query);
+    final StateQueryResult<KeyValueIterator<Windowed<GenericKey>, GenericRow>> result =
+        stateStore.getKafkaStreams().query(request);
+
+    try (KeyValueIterator<Windowed<GenericKey>, GenericRow> it =
+             result.getGlobalResult().getResult()) {
```

Comment:
We won't be querying a global store, so there won't be a global result. Looking through the code from the old cache bypass logic, I think you'll want to do something like this `io.confluent.ksql.execution.streams.materialization.ks.SessionStoreCacheBypass#findFirstNonEmptyIterator` on the PartitionResults instead.

Reply:
Ok, makes sense, changed that part, let me know what you think.